README.md
mypy-protobuf: Generate mypy stub files from protobuf specs [![Build Status](https://travis-ci.org/dropbox/mypy-protobuf.svg?branch=master)](https://travis-ci.org/dropbox/mypy-protobuf)
===========================================================

## Requirements
protoc 3.0.0 or greater  
python 2.7, 3.5, 3.6, 3.7 3.8
[python-protobuf >= 3.11.3](https://pypi.org/project/protobuf/3.7.1/)  
[mypy >= v0.790](https://pypi.org/project/mypy)

Other configurations may work, but are not supported in testing currently. We would be open to expanding this list if a need arises - file an issue on the issue tracker.

## Python Implementation
There is a python implementation of the plugin in `python/protoc-gen-mypy`. On windows
you will have to use `python/protoc_gen_mypy.bat` for the executable.

The plugin can be installed with
```
pip install mypy-protobuf
```
On posix, ensure that the protoc-gen-mypy script installed onto your $PATH. Then run.
```
protoc --python_out=output/location --mypy_out=output/location
```
Alternately, you can explicitly provide the path:
```
protoc --plugin=protoc-gen-mypy=path/to/protoc-gen-mypy --python_out=output/location --mypy_out=output/location
```
On windows, provide the bat file:
```
protoc --plugin=protoc-gen-mypy=path/to/protoc_gen_mypy.bat --python_out=output/location --mypy_out=output/location
```
To suppress output, you can run
```
protoc --python_out=output/location --mypy_out=quiet:output/location
```

## Go Implementation
There is a go implementation of the plugin in `go/src/protoc-gen-mypy`.
The import sort order can be customized to split between stdlib and project protos
by changing the `project` const at the top of the file (we use dropbox since our
proto files are namespaced under dropbox/)

To build the plugin: `go get github.com/dropbox/mypy-protobuf/go/src/protoc-gen-mypy`.

The plugin can be used by adding the built target to the command line
when running `protoc` (in addition to the normal plugin for output languages).

## Support
Dropbox internally uses both implementations. We internally directly use the python implementation.
However, the go implementation here is a code drop from the Dropbox internal implementation with periodic
re-upstreams. As a result, the python implementation will get more timely support. We encourage community
contribution to improve quality/testing to bring both implementations to parity.

## Contributing
Contributions to the implementations are welcome. Please run tests using `./run_test.sh`.
Ensure code is formatted using black.
```
pip3 install black
black python/ test/
```

## Contributors

### Dropboxers
- [@nipunn1313](https://github.com/nipunn1313)
- [@dzbarsky](https://github.com/dzbarsky)
- [@gvanrossum](https://github.com/gvanrossum)
- [@peterlvilim](https://github.com/peterlvilim)
- [@msullivan](https://github.com/msullivan)
- [@bradenaw](https://github.com/bradenaw)

### Others
- [@Ketouem](https://github.com/Ketouem)
- [@nmiculinic](https://github.com/nmiculinic)
- [@onto](https://github.com/onto)
- [@jcppkkk](https://github.com/jcppkkk)
- [@drather19](https://github.com/drather19)
- [@smessmer](https://github.com/smessmer)
- [@pcorpet](https://github.com/pcorpet)
- [@zozoens31](https://github.com/zozoens31)
- [@abhishekrb19](https://github.com/abhishekrb19)
- [@jaens](https://github.com/jaens)
- [@arussellsaw](https://github.com/arussellsaw)
- [@shabbyrobe](https://github.com/shabbyrobe)
- [@reorx](https://github.com/reorx)
- [@zifter](https://github.com/zifter)
- [@juzna](https://github.com/juzna)
- [@mikolajz](https://github.com/mikolajz)
- [@chadrik](https://github.com/chadrik)
- [@chrislawlor](https://github.com/chrislawlor)


Licence etc.
------------

1. License: Apache 2.0.
2. Copyright attribution: Copyright (c) 2017 Dropbox, Inc.
3. External contributions to the project should be subject to
   Dropbox's Contributor License Agreement (CLA):
   https://opensource.dropbox.com/cla/


CHANGELOG.md
## Upcoming

- Support ScalarMap and MessageMap generated types for map types in proto.
This will allow us to support `get_or_create`

```
message Message {
    map<int32, OuterMessage3> map_message = 17
}
```
and
```
message.map_message.get_or_create(0)
```

Before (1.23)
```
main.py:4: error: "MutableMapping[str, Nested]" has no attribute "get_or_create"  [attr-defined]
```
After (1.24) - there is no error

## 1.23

- Inherit FromString from superclass Message - rather than re-generating here. Fixes bug
in python2 usage `google/protobuf/type_pb2.pyi:92: error: Argument 1 of "FromString" is incompatible with supertype "Message"; supertype defines the argument type as "ByteString"  [override]`

## 1.22

- Update tested/required mypy version to 0.780 (picks up new typeshed annotations). Includes improved typing/error messages on Message.

Before (mypy < 0.780):
```
test_negative/negative.py:26: error: Argument 1 to "CopyFrom" of "Message" has incompatible type "str"; expected "Message"
```
After (mypy >= 0.780:
```
test_negative/negative.py:26: error: Argument 1 to "CopyFrom" of "Message" has incompatible type "str"; expected "Simple1"
```
- Update generated EnumTypeWrapper to be instances of EnumTypeWrapper - for more consistency
with generated python code. Most caller code should not require mypy type changes. Egh
`ProtoEnum.Value('first')` should work either way.

Generated Before (in 1.21)
```
class ProtoEnum(object):
    @classmethod
    def Value(cls, name: str) -> ProtoEnumValue
```

Generated After (in 1.22)
```
ProtoEnum: _ProtoEnum
class _ProtoEnum(google.protobuf.EnumTypeWrapper):
    def Value(self, name: str) -> ProtoEnumValue
```
- Remove autogenerated EnumTypeWrapper methods that are redundant to the typeshed parent class. Added testing for these.

## 1.21

- Support for module descriptor.
- Update mangling from `global__` to `message__`
- Fix bug in message typing for nested enums. Split EnumValue from EnumTypeWrapper. Enforces that constructing
an enum value must happen via a NewType wrapper to the int.

Example:
```
enum ProtoEnum {
    FIRST = 1;
    SECOND = 2;
}

mesage ProtoMsg {
    ProtoEnum enum = 1;
}
```
Generated Before (in 1.20):
```
class ProtoEnum(object):
    @classmethod
    def Value(cls, name: str) -> ProtoEnum

class ProtoMsg(Message):
    def __init__(self, enum: ProtoEnum) -> None
```
Generated After (in 1.21):
```
ProtoEnumValue = NewType('ProtoEnumValue', int)
class ProtoEnum(object):
    @classmethod
    def Value(cls, name: str) -> ProtoEnumValue
    
class ProtoMsg(Message):
    def __init__(self, enum: ProtoEnumValue) -> None
```
Migration Guide (with example calling code)

Before (with 1.20)
```
from msg_pb2 import ProtoEnum, ProtoMsg

def make_proto_msg(enum: ProtoEnum) -> ProtoMsg:
    return ProtoMsg(enum)
make_proto_msg(ProtoMsg.FIRST)
```
After (with 1.21)
```
from msg_pb2 import ProtoEnum, ProtoMsg

def make_proto_msg(enum: 'msg_pb2.ProtoEnumValue') -> ProtoMsg:
    return ProtoMsg(enum)
make_proto_msg(ProtoMsg.FIRST)
```

- Use inline-style rather than comment-style typing in the pyi file
- Remove MergeFrom/CopyFrom from generated code as it is in the Message superclass

## 1.20

- Black code formatting
- Fix message/field name aliasing when field name matches a message/enum name

## 1.19

- Allow omitting required proto2 fields from constructor parameters
- Support and testing for python 3.8
- Support for python-protobuf to 3.11.3

## 1.18

- Use `entry_points:console_scripts` to support long paths to the python interpreter

## 1.17

- Update to newer mypy version - including minor changes to typeshed

## 1.16

- Absolute path to necessary python
- Add forward reference string literal support for enums
- Alias builtin types to avoid collision with field name

## 1.15

- Add `class` to set of python keywords

## 1.14

- Add `Message.DESCRIPTOR`


## Older changelogs not available. Check git log if you need them!


test/test_generated_mypy.py
"""
This file serves two purposes
1) Run a test which ensures that mypy is producing the expected output. For
developer convenience, in the case which it doesn't, it'll copy over the
output to the expected output, so you can view the output diff in your
code reviews. This will fail in CI if there is any inconsistency.

2) This is a file which should mypy with success. See test_negative for
a file that should have failures.
"""

import glob
import os
import pytest
import six

import test.proto.test_pb2 as test_pb2
from test.proto.test_pb2 import DESCRIPTOR, FOO, OuterEnum, Simple1, Simple2
from test.proto.test3_pb2 import SimpleProto3, OuterMessage3
from test.proto.Capitalized.Capitalized_pb2 import lower, lower2, Upper

from typing import Any

MYPY = False
if MYPY:
    from test.proto.test_pb2 import OuterEnumValue


def _is_summary(l):
    # type: (str) -> bool
    """Checks if the line is the summary line 'Found X errors in Y files (checked Z source files)'"""
    return l.startswith("Found ") and l.endswith("source files)\n")


def test_generate_mypy_matches():
    # type: () -> None
    proto_files = glob.glob("proto/test/proto/*.proto") + glob.glob(
        "proto/test/proto/*/*.proto"
    )
    assert len(proto_files) == 9  # Just a sanity check that all the files show up

    failures = []
    for fn in proto_files:
        assert fn.endswith(".proto")
        fn_split = fn.split(os.sep)  # Eg. [proto, test, proto, test.com, test.proto]
        assert fn_split[-1].endswith(".proto")
        last = fn_split[-1][: -len(".proto")] + "_pb2.pyi"  # Eg [test_pb2.proto]
        components = fn_split[1:-1]  # Eg. [test, proto, test.com]
        components = [
            c.replace(".", os.sep) for c in components
        ]  # Eg. [test, proto, test/com]
        components.append(last)  # Eg. [test, proto, test/com, test_pb2.proto]

        output = os.path.join(*components)

        components[
            -1
        ] += ".expected"  # Eg [test, proto, test/com, test_pb2.proto.expected]

        expected = os.path.join(*components)

        assert os.path.exists(output)

        output_contents = open(output).read()
        expected_contents = open(expected).read() if os.path.exists(expected) else None

        if output_contents != expected_contents:
            open(expected, "w").write(output_contents)
            failures.append(
                (
                    "%s doesn't match %s. This test will copy it over. Please rerun"
                    % (output, expected)
                )
            )

    if failures:
        raise Exception(str(failures))


def test_generate_negative_matches():
    # type: () -> None
    """Confirm that the test_negative expected file matches an error for each line"""
    test_negative_lines = open("test_negative/negative.py").readlines()
    # Grab the line number of the failures
    errors_27 = set(
        int(l.split(":")[1])
        for l in open("test_negative/output.expected.2.7").readlines()
        if not _is_summary(l)
    )
    errors_35 = set(
        int(l.split(":")[1])
        for l in open("test_negative/output.expected.3.5").readlines()
        if not _is_summary(l)
    )

    expected_errors_27 = set(
        idx + 1 for idx, line in enumerate(test_negative_lines) if "E:2.7" in line
    )
    expected_errors_35 = set(
        idx + 1 for idx, line in enumerate(test_negative_lines) if "E:3.5" in line
    )

    assert errors_27 == expected_errors_27
    assert errors_35 == expected_errors_35

    # Some sanity checks to make sure we don't mess this up. Please update as necessary.
    assert len(errors_27) == 36
    assert len(errors_35) == 36


def test_func():
    # type: () -> None
    s = Simple1(a_string="hello")

    s = Simple1()
    s.a_string = "Hello"
    s.a_repeated_string.append("Hello")

    s2 = Simple1.FromString(s.SerializeToString())
    assert s2.a_string == "Hello"

    s3 = Simple1()
    s3.ParseFromString(s.SerializeToString())
    assert s3.a_string == "Hello"

    s4 = Simple1()
    s4.CopyFrom(s)
    assert s4.a_string == "Hello"

    l = lower2()
    l.upper.Lower.a = 2
    assert l == lower2(upper=Upper(Lower=lower(a=2)))


def test_enum():
    # type: () -> None
    e = FOO
    e = OuterEnum.Value("BAR")
    assert OuterEnum.Value("BAR") == 2
    assert OuterEnum.Name(e) == "BAR"
    assert OuterEnum.keys() == ["FOO", "BAR"]
    assert OuterEnum.values() == [1, 2]
    assert OuterEnum.items() == [("FOO", 1), ("BAR", 2)]

    # Make sure we can assure typing with a couple of techniques
    e2 = OuterEnum.Value("BAR")  # type: test_pb2.OuterEnumValue
    assert OuterEnum.Name(e2) == "BAR"
    e3 = OuterEnum.Value("BAR")  # type: OuterEnumValue
    assert OuterEnum.Name(e3) == "BAR"
    e4 = OuterEnum.Value("BAR")  # type: int
    assert OuterEnum.Name(e2) == "BAR"


def test_has_field_proto2():
    # type: () -> None
    """For HasField which is typed with Literal"""
    s = Simple1()
    s.a_string = "Hello"

    # Proto2 tests
    assert s.HasField(u"a_string")
    assert s.HasField("a_string")
    if six.PY2:
        assert s.HasField(b"a_string")
    assert not s.HasField("a_inner")
    assert not s.HasField("a_enum")
    assert not s.HasField("a_oneof")

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(
        ValueError, match="Protocol message Simple1 has no field garbage."
    ):
        s_untyped.HasField("garbage")
    with pytest.raises(
        ValueError,
        match='Protocol message Simple1 has no singular "a_repeated_string" field',
    ):
        s_untyped.HasField("a_repeated_string")
    if six.PY3:
        with pytest.raises(TypeError, match="bad argument type for built-in operation"):
            s_untyped.HasField(b"a_string")


def test_has_field_proto3():
    # type: () -> None
    s = SimpleProto3()
    assert not s.HasField(u"outer_message")
    assert not s.HasField("outer_message")
    if six.PY2:
        assert not s.HasField(b"outer_message")
    assert not s.HasField("a_oneof")

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(
        ValueError, match="Protocol message SimpleProto3 has no field garbage."
    ):
        s_untyped.HasField(u"garbage")
    with pytest.raises(
        ValueError,
        match='Can\'t test non-submessage field "SimpleProto3.a_string" for presence in proto3.',
    ):
        s_untyped.HasField(u"a_string")
    with pytest.raises(
        ValueError,
        match='Can\'t test non-submessage field "SimpleProto3.a_outer_enum" for presence in proto3.',
    ):
        s_untyped.HasField("a_outer_enum")
    with pytest.raises(
        ValueError,
        match='Protocol message SimpleProto3 has no singular "a_repeated_string" field',
    ):
        s_untyped.HasField(u"a_repeated_string")
    if six.PY3:
        with pytest.raises(TypeError, match="bad argument type for built-in operation"):
            s_untyped.HasField(b"outer_message")


def test_clear_field_proto2():
    # type: () -> None
    """For ClearField which is typed with Literal"""
    s = Simple1()
    s.a_string = "Hello"

    # Proto2 tests
    s.ClearField(u"a_string")
    if six.PY2:
        s.ClearField(b"a_string")
    s.ClearField("a_string")
    s.ClearField("a_inner")
    s.ClearField("a_repeated_string")
    s.ClearField("a_oneof")
    s.ClearField(b"a_string")

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(ValueError, match='Protocol message has no "garbage" field.'):
        s_untyped.ClearField("garbage")


def test_clear_field_proto3():
    # type: () -> None
    """For ClearField which is typed with Literal"""
    s = SimpleProto3()
    s.a_string = "Hello"

    # Proto2 tests
    s.ClearField(u"a_string")
    if six.PY2:
        s.ClearField(b"a_string")
    s.ClearField("a_string")
    s.ClearField("a_outer_enum")
    s.ClearField("outer_message")
    s.ClearField("a_repeated_string")
    s.ClearField("a_oneof")
    s.ClearField(b"a_string")

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(ValueError, match='Protocol message has no "garbage" field.'):
        s_untyped.ClearField("garbage")


def test_which_oneof_proto2():
    # type: () -> None
    s = Simple1()

    assert s.WhichOneof("a_oneof") is None
    s.a_oneof_1 = "hello"
    assert s.WhichOneof("a_oneof") == "a_oneof_1"
    assert s.WhichOneof(u"a_oneof") == "a_oneof_1"
    assert s.WhichOneof(b"a_oneof") == "a_oneof_1"
    assert type(s.WhichOneof("a_oneof")) == str
    assert s.HasField(s.WhichOneof("a_oneof"))

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(
        ValueError, match='Protocol message has no oneof "garbage" field.'
    ):
        s_untyped.WhichOneof("garbage")


def test_which_oneof_proto3():
    # type: () -> None
    s = SimpleProto3()

    assert s.WhichOneof("a_oneof") is None
    s.a_oneof_1 = "hello"
    s.b_oneof_1 = "world"
    assert s.WhichOneof("a_oneof") == "a_oneof_1"
    assert s.WhichOneof(u"a_oneof") == "a_oneof_1"
    assert s.WhichOneof(b"a_oneof") == "a_oneof_1"
    assert type(s.WhichOneof("a_oneof")) == str
    assert s.HasField(s.WhichOneof("a_oneof"))
    assert s.HasField(s.WhichOneof("b_oneof"))

    # Erase the types to verify that incorrect inputs fail at runtime
    # Each test here should be duplicated in test_negative to ensure mypy fails it too
    s_untyped = s  # type: Any
    with pytest.raises(
        ValueError, match='Protocol message has no oneof "garbage" field.'
    ):
        s_untyped.WhichOneof("garbage")


def test_constructor_proto2():
    # type: () -> None
    x = Simple2()  # It's OK to omit a required field from the constructor.
    assert not x.HasField("a_string")

    x = Simple2(a_string=None)  # It's OK to pass None for a required field.
    assert not x.HasField("a_string")


def test_message_descriptor_proto2():
    # type: () -> None
    assert Simple1().DESCRIPTOR.full_name == "test.Simple1"
    assert Simple1.DESCRIPTOR.full_name == "test.Simple1"


def test_message_descriptor_proto3():
    # type: () -> None
    assert SimpleProto3().DESCRIPTOR.full_name == "test3.SimpleProto3"
    assert SimpleProto3.DESCRIPTOR.full_name == "test3.SimpleProto3"


def test_enum_descriptor():
    # type: () -> None
    assert OuterEnum.DESCRIPTOR.name == "OuterEnum"


def test_module_descriptor():
    # type: () -> None
    assert DESCRIPTOR.name == "test/proto/test.proto"

def test_mapping_type():
    # type: () -> None
    s = SimpleProto3()
    s.map_scalar[5] = "abcd"
    assert s.map_scalar[5] == "abcd"

    s.map_message[5].a_bool = True
    assert s.map_message[5] == OuterMessage3(a_bool=True)

    assert s.map_message.get_or_create(6) == OuterMessage3()
    assert s.map_message[6] == OuterMessage3()
    assert s.map_message.get_or_create(6) == OuterMessage3()


test_negative/negative.py
"""
This code is intended to have mypy failures which we will ensure
show up in the output.
"""

import six

from typing import (
    List,
    Text,
)

from test.proto.test_pb2 import DESCRIPTOR, FOO, Simple1, Simple2
from test.proto.test3_pb2 import OuterEnum, OuterEnumValue, SimpleProto3
from test.proto.dot.com.test_pb2 import TestMessage

s = Simple1()
s.a_string = "Hello"

s2 = Simple1.FromString(s.SerializeToStringg())  # E:2.7 E:3.5

s3 = Simple1()
s3.ParseFromString(s)  # E:2.7 E:3.5

s4 = Simple1()
s4.CopyFrom(s.SerializeToString())  # E:2.7 E:3.5

s5 = Simple1()
l = []  # type: List[int]
l.extend(s5.a_repeated_string)  # E:2.7 E:3.5

tm = TestMessage(foo=55)  # E:2.7 E:3.5

e = FOO
e = 3  # E:2.7 E:3.5

# Proto2
s.HasField("garbage")  # E:2.7 E:3.5
s.HasField("a_repeated_string")  # E:2.7 E:3.5

# Proto3
s6 = SimpleProto3()
s6.HasField(u"garbage")  # E:2.7 E:3.5
s6.HasField(u"a_string")  # E:2.7 E:3.5
s6.HasField(u"outer_enum")  # E:2.7 E:3.5
s6.HasField(u"a_repeated_string")  # E:2.7 E:3.5

# Proto2
s.ClearField("garbage")  # E:2.7 E:3.5

# Proto3
s6.ClearField("garbage")  # E:2.7 E:3.5

# Proto2 WhichOneof
s.WhichOneof("garbage")  # E:2.7 E:3.5
a = 5
a = s.WhichOneof("a_oneof")  # E:2.7 E:3.5
b = s.WhichOneof("a_oneof")
s.HasField(b)  # allowed
simple2 = Simple2(a_string="abc")
simple2.HasField(b)  # E:2.7 E:3.5

# Proto3 WhichOneof
s6.WhichOneof("garbage")  # E:2.7 E:3.5
a3 = 5
a3 = s6.WhichOneof("a_oneof")  # E:2.7 E:3.5
b3 = s6.WhichOneof("a_oneof")
s6.HasField(b3)  # allowed
simple2.HasField(b3)  # E:2.7 E:3.5  - it's a text but not one of the literals

# Overload WhichOneof
c = s6.WhichOneof("a_oneof")
c = s6.WhichOneof("b_oneof")  # E:2.7 E:3.5

# Message DESCRIPTOR should detect invalid access via instance or class:
SimpleProto3.DESCRIPTOR.Garbage()  # E:2.7 E:3.5
SimpleProto3().DESCRIPTOR.Garbage()  # E:2.7 E:3.5

# Enum DESCRIPTOR should detect invalid access:
OuterEnum.DESCRIPTOR.Garbage()  # E:2.7 E:3.5

# Module DESCRIPTOR should detect invalid access:
DESCRIPTOR.Garbage()  # E:2.7 E:3.5

# Enum value type should be an EnumValueType convertible to int
enum = OuterEnumValue(5)
enum = s6.a_outer_enum
as_int = 5
as_int = s6.a_outer_enum
s6.a_outer_enum.FOO  # E:2.7 E:3.5

# Name/Value/items should inherit value type from _EnumTypeWrapper
OuterEnum.Name(OuterEnumValue(5))
OuterEnum.Name(5)  # E:2.7 E:3.5
OuterEnum.Name(OuterEnum.Value("BAR"))
OuterEnum.Name(SimpleProto3.InnerEnum.Value("BAR"))  # E:2.7 E:3.5
OuterEnum.Name(OuterEnum.values()[0])
OuterEnum.Name(SimpleProto3.InnerEnum.values()[0])  # E:2.7 E:3.5
OuterEnum.Name(OuterEnum.items()[0][1])
OuterEnum.Name(SimpleProto3.InnerEnum.items()[0][1])  # E:2.7 E:3.5

# Map field does not have get_or_create when mapping to a scalar type
s7 = SimpleProto3()
s7.map_scalar.get_or_create(0)  # E:2.7 E:3.5
# Incorrect key type should error
s7.map_scalar.get_or_create("abcd")  # E:2.7 E:3.5
s7.map_scalar.get("abcd")  # E:2.7 E:3.5
s7.map_message.get("abcd")  # E:2.7 E:3.5
# Incorrect value type should error
map_val = 5
map_val = s7.map_scalar.get(0)  # E:2.7 E:3.5
map_val = s7.map_message.get(0)  # E:2.7 E:3.5
